### PR TITLE
Use Java 8 APIs.

### PR DESCRIPTION
**Pull request by Isira-Seneviratne**
#### What is it?
- [ ] Bug fix (user facing)
- [ ] Feature (user facing)
- [x] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
* Add support for [Java 8 API desugaring](https://developer.android.com/studio/write/java8-support#library-desugaring).
* Make use of APIs introduced in Java 8, such as the `Comparator.comparing` methods (using the [the Java 8 Date/Time API](https://developer.android.com/reference/java/time/package-summary) has been moved to [a separate PR](https://github.com/TeamNewPipe/NewPipe/pull/4556)).
* Using Java 8 API desugaring also incorporates [bug fixes](https://bugs.openjdk.java.net/browse/JDK-8028564) for `ConcurrentHashMap` on Android Lollipop.

#### Testing apk

[debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5470477/debug.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

### CODE REVIEW

**Comment by wb9688** [comment]
Please squash some commits, so that all Date/Time related changes are one commit, all removeIf() related changes are one commit, etc.

**Comment by Isira-Seneviratne** [comment]
> 
> 
> Please squash some commits, so that all Date/Time related changes are one commit, all removeIf() related changes are one commit, etc.

Okay.

> 
> Btw is there any reason for not using a newer Gradle version?

No, it was set to that specific version when updating the Android Gradle plugin.

File: app/build.gradle [hunk 1]
```diff
@@ -139,6 +143,8 @@ afterEvaluate {
 }
 
 dependencies {
+    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9'
```
Comment (opusforlife2) [inline app/build.gradle:146]: Any harm in using the latest version instead?

Comment (Isira-Seneviratne) [reply app/build.gradle:146]: I'll update it, thanks.

Comment (Isira-Seneviratne) [reply app/build.gradle:146]: On that note, should I also update `debug.zip`?

Comment (opusforlife2) [reply app/build.gradle]: Leave it for the next release cycle, I think.

**Comment by Isira-Seneviratne** [comment]
@Stypox I'm merging the changes from the master branch.

**Comment by Isira-Seneviratne** [comment]
I fixed a merge conflict.

**Review by Stypox — CHANGES_REQUESTED** [review]
Please undo the changes to build.gradle and gradle/wrapper/gradle-wrapper.properties, since the grandle update is covered by #3509

**Comment by wb9688** [comment]
@Stypox: It's not possible to have these improvements without those changes

**Comment by Stypox** [comment]
@wb9688 I know, I just keep those files modified locally without pushing them 😉

**Comment by Isira-Seneviratne** [comment]
> @wb9688 I know, I just keep those files modified locally without pushing them 😉

The thing is, the Gradle plugin version in the dev branch is currently 3.6.3, but API desugaring requires at least 4.0.0.

**Comment by Stypox** [comment]
@Isira-Seneviratne could you upload a new apk? After that I will review, test and merge

**Comment by Isira-Seneviratne** [comment]
Sure.

**Comment by Isira-Seneviratne** [comment]
I uploaded a new APK.

**Review by Stypox — APPROVED** [review]
Thank you! Code looks good. I tested all the activities you changed on Android 4.4, 7.0 and 10, and everything worked fine. :-D

**Comment by Stypox** [comment]
@TobiGr merge (now GitHub waits indefinitely for `synk` to pass, but since you disabled the webhook it won't ever report a status)

**Pull request merged**